feat(laverna): create laverna

boneskull opened the pull request:
# Laverna

 ![image](https://github.com/LavaMoat/LavaMoat/assets/924465/c4ee9f22-fac6-4c16-8340-da660c64bccb)

> Publish multiple workspaces (that's all)

If you're familiar with [lerna](https://lerna.js.org), **Laverna** does this: `lerna publish from-package`.

If you're unfamiliar with it: **Laverna** publishes all workspaces that haven't yet been published.

## Features

**Laverna** is a thin wrapper around `npm publish` which:

- Invokes `npm publish` for all workspaces _that haven't yet been published_
- Handles new packages (via a flag)
- Ignores private packages
- Requires a confirmation (by default)
- Provides a "dry-run" mode
- Prints all output from `npm publish`, including the output of `npm pack`, to enable review

## Non-Features

Perhaps more importantly, **Laverna**:

- Builds nothing
- Bumps no versions
- Does not write to `package.json`
- Does not interact with `git` (no tags, no commits, no pushes)
- Does not interact with GitHub (no releases)
- Does not cache anything

## Supported Environments

- Node.js v18.13.0+
- `npm` v8.19.3+

## Install

It's recommended to install as a dev dependency:

```shell
npm install @lavamoat/laverna -D
```

## Usage Summary

```plain
laverna [options..]

Publish multiple workspaces (that's all)

Options:

 --dryRun        - Enable dry-run mode
 --root=<path>   - Path to workspace root (default: current working dir)
 --newPkg=<name> - Workspace <name> has never been published (repeatable)
 --yes/-y        - Skip confirmation prompt

Problems? Visit https://github.com/LavaMoat/lavamoat/issues
```

## Examples

### Using Release Please or Similar

For a typical release workflow, you might:

1. Bump workspace versions (e.g. via [release-please-action](https://github.com/google-github-actions/release-please-action))
2. In your updated working copy's workspace root, run `npm exec laverna` (or `npm exec laverna -- --dry-run` first)
3. Bask in glory

### Publishing a New Package

If you're publishing a package in a new workspace, you might:

1. Set the initial version of the new package (call it `foo`) in its `package.json`.
2. Run `npm exec laverna -- --newPkg=foo` (or `npm exec laverna -- --newPkg=foo --dry-run` first)
3. Profit

### Automating Publishes

You'll need to figure out how to add `--newPkg` to your CI/CD pipeline. Otherwise, use the `--yes` flag when invoking `laverna`.

## API

**Laverna** provides a programmable JavaScript API (for what it's worth).

Refer to the TypeScript definitions for details.

## Notes

[LavaMoat](https://github.com/LavaMoat/LavaMoat) needed something that did this, but didn't need Lerna for anything else. So, here we are.

## License

Copyright © 2023 Consensys, Inc. Licensed MIT

boneskull committed:
feat(laverna): create laverna

This copies `scripts/publish.js` into its own package named `@lavamoat/laverna`.

boneskull committed:
chore(laverna): add some more e2e tests

boneskull committed:
chore(laverna): add some keywords

boneskull committed:
chore(laverna): attempt to fix tests

legobeat commented on packages/laverna/README.md:
```diff
@@ -0,0 +1,95 @@
+
+Options:
+
+ --dryRun        - Enable dry-run mode
```
While using `--camelCasedFlags` is consistent with the rest of lavamoat cli tooling, recognizing `--dry-run` and `--new-pkg` would make transitioning more seamless. 

LGTM otherwise!

boneskull replied:
the old publish script also uses camelCased flags, so this wouldn't be an impediment there.

I _would_ like to use kebab-cased flags, but I'm using Node.js' parsing facilities, and if I want an object that gives me both, I'd need to code it myself.

legobeat replied:
In any case, the examples below still give the non-recognized form as examples, so one of them needs to change

https://github.com/LavaMoat/LavaMoat/pull/836#discussion_r1457660222

legobeat commented on packages/laverna/README.md:
```diff
@@ -0,0 +1,95 @@
+If you're publishing a package in a new workspace, you might:
+
+1. Set the initial version of the new package (call it `foo`) in its `package.json`.
+2. Run `npm exec laverna -- --newPkg=foo` (or `npm exec laverna -- --newPkg=foo --dry-run` first)
```
`--dry-run` results in `ERR_PARSE_ARGS_UNKNOWN_OPTION`. I think it should support the flag in this format (either in addition to or instead of the current one). I'd certainly keep mixing it up, for one.

boneskull committed:
chore(laverna): handle kebab-case flags

This adds support for `--new-pkg` and `--dry-run`, though it remains undocumented.

If someone uses both `--new-pkg` and `--newPkg` and/or `--dry-run` and `--dryRun`, Laverna will exit with an error.

legobeat commented on packages/laverna/src/cli.js:
```diff
@@ -0,0 +1,110 @@
+
+    // prefer camelCase
+    opts.dryRun = opts.dryRun ?? opts['dry-run']
+    opts.newPkg = opts.newPkg ?? opts['new-pkg']
```
```suggestion
    // prefer camelCase
    opts.dryRun = opts.dryRun || opts['dry-run']
    opts.newPkg = new Set([..opts.newPkg, ...opts['new-pkg'])
```

- `--newPkg` / `--new-pkg` is repeatable. While a bit silly, I don't see why `--newPkg=foo --new-pkg=bar` should not be considered valid?

- being a boolean flag defaulting to `false` and not accepting passed values/negation, `--dryRun`, `--dry-run`, and `--dryRun --dry-run` should all be redundant/equivalent



So in both cases we don't really have notion of conflicts and need for precedence, so could prob skip these checks alltogether?

boneskull replied:
I don't want to write code to merge and dedupe the arrays generated by passing `--newPkg` and `--new-pkg`.  I also didn't want inconsistent behavior around `--dryRun`.  

I didn't really want to write _any_ code to handle both kebab-case and camelCase.  If we need stuff like that, I'd rather just pull in `yargs`.

I'm inclined to rip the kebab-case stuff out.

boneskull replied:
Another way to look at it: we can add all the bells & whistles later, if desired.  While I absolutely understand the conventions and UX around CLI apps, it's not my intent to give heavy weight usability (for now)--especially since this is something nobody else asked for.

legobeat replied:
> I don't want to write code to merge and dedupe the arrays generated by passing --newPkg and --new-pkg


Well, it's there in the suggestion now; I don't think you'll need to write any new code for that? 


> I didn't really want to write any code to handle both kebab-case and camelCase

The suggestion here is indeed to reduce the special-handling - as it doesn't seem necessary.

>  I also didn't want inconsistent behavior around `--dryRun`.

What's the inconsistent behavior?

legobeat commented on packages/laverna/src/cli.js:
```diff
@@ -0,0 +1,110 @@
+  'new-pkg': {
+    type: 'string',
+    multiple: true,
+  },
```
```suggestion
  newPkg: {
    type: 'string',
    multiple: true,
    default: [],
  },
  'new-pkg': {
    type: 'string',
    multiple: true,
    default: [],
  },
```

legobeat commented on packages/laverna/src/defaults.js:
```diff
@@ -0,0 +1,67 @@
+    dryRun: false,
+    yes: false,
+    root: DEFAULT_ROOT,
+    newPkg: [],
```
```suggestion
    newPkg: new Set(),
```

boneskull replied:
since `parseArgs` won't give me a `Set` back, I'll need to create more types. Given the expected size of the `newPkg` array, this seems like premature optimization--and it really does not matter if the same package occurs multiple times in the array, since we do not iterate over it.

legobeat commented on packages/laverna/src/laverna.js:
```diff
@@ -0,0 +1,415 @@
+      if (/** @type {NodeJS.ErrnoException} */ (e).code === 'ENOENT') {
+        throw new Error(`Could not find npm: ${e}`)
+      }
+      if (newPkg.includes(name)) {
```
```suggestion
      if (newPkg.has(name)) {
```

legobeat reviewed (dismissed):
TS errors in tests:

```
Error: packages/laverna/test/laverna.spec.js(24,44): error TS2694: Namespace '"node:test"' has no exported member 'Mock'.
Error: packages/laverna/test/laverna.spec.js(25,35): error TS2694: Namespace '"node:test"' has no exported member 'Mock'.
Error: packages/laverna/test/laverna.spec.js(419,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(422,16): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(506,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(508,21): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(530,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(532,21): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(693,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(696,16): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(756,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(758,21): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(788,6): error TS7006: Parameter 'call' implicitly has an 'any' type.
Error: packages/laverna/test/laverna.spec.js(791,16): error TS7006: Parameter 'arg' implicitly has an 'any' type.
Error: Process completed with exit code 2.
```

legobeat commented on packages/laverna/package.json:
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@lavamoat/laverna",
+  "version": "0.0.0",
+  "description": "Publish multiple workspaces (that's all)",
```
```suggestion
  "description": "Publish multiple NPM workspaces (that's all)",
```

boneskull committed:
chore: fix invalid tag order

This is a workaround for https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/226

boneskull committed:
chore(laverna): peer review changes

- `--dry-run` and `--dryRun` are both supported w/o conflict
- `--newPkg` and `--new-pkg` are both supported w/o conflict

boneskull committed:
chore: vendor some Mock types

legobeat commented on package.json:
```diff
@@ -30,9 +30,8 @@
     "release:publish": "node scripts/publish.js --newPkg=${npm_config_newpkg}",
     "release:rebuild": "npm run rebuild && npm run test:prep && npm run test:workspaces",
     "setup": "husky install; npm run rebuild",
-    "test": "npm run build && npm run test:prep && npm run test:workspaces && npm run test:scripts",
```
The `scripts/publish.js` is still used by the LavaMoat project for package publishing - it looks premature to remove tests scripts and disable tests in CI as long as that is the case?

Seems like this should be part of #903

boneskull replied:
yes, that looks like I missed it

boneskull committed:
chore(deps): update lockfile for laverna

boneskull committed:
chore(laverna): update default values in CLI

Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>

naugtur commented on packages/laverna/ava.config.js:
```diff
@@ -0,0 +1,13 @@
+try {
+  if (require('node:test').mock) {
+    config.files.push('test/laverna.spec.js')
+  }
```
I like this solution. Would be nice to make this more explicit.
```suggestion
  } else {
    console.error('[SKIP] Laverna tests skipped - too old Node.js version')
  }
```

naugtur commented on package.json:
```diff
@@ -96,6 +96,7 @@
   "prettier": {
     "jsdocCommentLineStrategy": "keep",
     "jsdocPreferCodeFences": true,
+    "jsdocTagsOrder": "{\"template\": 24.5}",
```
I don't understand this and therefore fear it. I'm a simple man. Please help me understand why we want this here.

boneskull replied:
heh, well, I would have explained myself, but it's `package.json`, so.

https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/226

naugtur commented on scripts/publish.js:
```diff
@@ -315,7 +339,13 @@ const Publish = {
+    summary: string
+    detail: string
+  }
+}} */
```
prettier went bonkers?

boneskull replied:
hmm, changes to this file should not be in this PR.  I'll pull that out

naugtur commented on packages/laverna/test/fixture/new-version/packages/lavamoat/package.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "lavamoat",
+  "version": "31337.420.69"
```
this random number generator, so mainstream

boneskull committed:
chore(laverna): notify when test skipped

Co-authored-by: Zbyszek Tenerowicz <naugtur@gmail.com>

boneskull committed:
chore(laverna): fix eslint config
